cloud/amazon/ec2_vpc_nat_gateway.py

```diff
@@ -0,0 +1,356 @@
+    except botocore.exceptions.ClientError as ce:
+        if ce.response['Error']['Code'] == "NatGatewayMalformed":
+            module.fail_json(msg="Invalid NAT Gateway ID")
+        else:
+            module.fail_json(msg=ce.message)
+
+
+def ensure_nat_gateway_present(ec2_client=None, module=None):
+    """ Ensure NAT gateway with specified parameters exists and call create if it does not """
+    # Check allocation-id refers to existing EIP allocation
+    allocation_id = module.params.get('allocation_id')
+    try:
+        ec2_client.describe_addresses(AllocationIds=[module.params.get('allocation_id')])
+    except botocore.exceptions.ClientError:
+        module.fail_json(msg="allocation: %s does not exist." % allocation_id)
+    # Check subnet exists
+    subnet_id = module.params.get('subnet_id')
+    try:
+        ec2_client.describe_subnets(SubnetIds=[subnet_id])
+    except botocore.exceptions.ClientError:
+        module.fail_json(msg="subnet: %s does not exist." % subnet_id)
+
+    # Get existing number of NAT Gateways in specified subnet
+    nat_gateways = ec2_client.describe_nat_gateways(Filters=[{'Name': 'subnet-id',
+                                                              'Values': [subnet_id]},
+                                                             {'Name': 'state',
+                                                              'Values': ['available', 'pending',
+                                                                         'deleting']}])
+    num_nat_gateways = len(nat_gateways.get('NatGateways')) if nat_gateways.get('NatGateways') else 0
+
+    # Check max_count has not been exceeded before attempting to create
+    if module.params.get('max_count') < num_nat_gateways:
+        module.fail_json(
+            msg="The maximum desired number of NAT Gateways specified in "
+                "max_count has already been exceeded.")
+    try:
+        gateway_status_list = get_nat_gateway_status_list(ec2_client=ec2_client, module=module)
+        for gateway_status in gateway_status_list:
+            if gateway_status.get('state') == 'available':
+                results = dict(changed=False, nat_gateway_id=gateway_status['nat_gateway_id'])
+                module.exit_json(**results)
+        else:
+            # Check creation of NAT Gateway does not exceed max_count
+            if module.params.get('max_count') == num_nat_gateways:
+                module.fail_json(
+                    msg="The maximum desired number of NAT Gateways specified in "
+                        "max_count (default: 1) has already been met.")
+            if not module.check_mode:
+                create_nat_gateway(ec2_client=ec2_client, module=module)
+            else:
+                module.exit_json(dict(change=True))
+    except botocore.exceptions.ClientError as ce:
+        if ce.response['Error']['Code'] == "NatGatewayLimitExceeded":
+            module.fail_json(msg="The NAT Gateway limit has been exceeded.")
```

Comment:
Prefer to go with boto3 message.  Not sure what it says though! Is it not good enough?

Better exception info as per guidelines:

`module.fail_json(msg=ce.message, **camel_dict_to_snake_dict(ce.response))`